053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+
+I propose that we comment out these up migrations (so future deployments don't
+get these tables/triggers), and leave any database initialization (or whatever
+appropriate backend setup) to `BuildEvent.Setup()`. This means `Setup` will
+have to be "smart" (i.e. have to handle the cases where the initialization has
+already been applied). It also makes it difficult to make changes to the
+schema, unlike with migrations - each `EventStore` (that stores data in a
+structured way - this may just end up being Postgres) will need to basically
+implement its own form of migrations if we need to change the schema. My guess
+is that we won't be changing the schema too often, though (given that we
+haven't changed `build_events` since the initial migration).
+
+
+## Configuration
+
+One thing I haven't really explored is how operators would configure the
+backend store. I think it could be similar to how Credential managers are
+configured - there are groups of independent configuration options for each
+possible `EventStore`. If certain flags are configured for a given backend, use
+that backend.
+
+
+# Open Questions
+
+* Does `Key` need to be `interface{}`, or can it be more specific like `uint`?
+  For backends that don't support auto-incrementing, but still need some
+  ordering (if the timestamp isn't sufficient), we **could** create a Postgres
+  sequence for every backend (external to the `EventStore`), and change the
+  signature to `Put(build db.Build, eventID uint, event atc.Event)`
+
+* Do we care about the `Last-Event-ID` header on the
+  `/api/v1/builds/{build_id}/events` endpoint? I don't think we use it anywhere
+  internally, but I guess we may want to keep it for backwards compatibility?
+  It's not exposed by `fly` or `go-concourse`, either.
+
+* Would any event stores benefit from bringing their own notification bus
+  implementation, rather than relying on Postgres' notification bus? Does using
+  the Postgres notification bus defeat the purpose of having an external store in
+  any way (e.g. could this be a bottleneck)?
+
+* Users won't be able to access build events that exist in the main Postgres
+  database if they switch to a new backend store. How can we make the
+  transition easier? One idea I have is to have a `FallbackEventStore` that
+  composes two `EventStores` - it will always `Put` to the primary one, but if
+  `Get` fails or returns no events from the primary `EventStore`, try the
+  fallback `EventStore`.
+
+
+# Answered Questions
+
+
+# New Implications
+
+* By default, there will be no change - Concourse will default to using a
+  Postgres `EventStore` using the same main Postgres database.
+* Operators will be able to configure a different store
+* If they configure a new store, they won't be able to access any existing
+  build events that are currently stored in Postgres without migrating them
+  over (how should we recommend doing this?)
```

Comment:
I think it should be a standalone tool, not part of the ATC. Otherwise you're forcing folks to upgrade to get it, or to get bugfixes to such a migration tool.